usr.bin/vgrind/vgrind.1
.\"	BSDI	$Id: vgrind.1,v 1.2 1994/01/03 20:47:09 polk Exp $
.\"
.\" Copyright (c) 1980 ,1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)vgrind.1	6.4 (Berkeley) 4/23/91
.\"
.Dd April 23, 1991
.Dt VGRIND 1
.Os BSD 4
.Sh NAME
.Nm vgrind
.Nd grind nice listings of programs
.Sh SYNOPSIS
.Nm vgrind
.Op Fl 
.Op Fl T Ar dev
.Op Fl d Ar file
.Op Fl f
.Op Fl h Ar header
.Op Fl l Ar language
.Op Fl n
.Op Fl sn
.Op Fl t
.Op Fl x
.Ar name Ar ...
.Sh DESCRIPTION
.Nm Vgrind
formats the program sources which are arguments 
in a nice style using
.Xr troff 1
Comments are placed in italics, keywords in bold face,
and the name of the current function is listed down the margin of each
page as it is encountered.
.Pp
.Nm Vgrind
runs in two basic modes, filter mode (see the
.Fl f
option) or regular mode.  In filter mode 
.Nm vgrind
acts as a filter in a manner similar to
.Xr tbl 1 .
The standard input is passed directly to the standard output except
for lines bracketed by the 
.Em troff-like
macros:
.Bl -tag -width Ds
.It \&.vS
starts processing
.It \&.vE
ends processing
.El
.Pp
These lines are formatted as described above.  The output from this
filter can be passed to 
.Xr troff
or
.Xr groff
for output.  There need be no particular ordering with 
.Xr eqn 1
or
.Xr tbl 1 .
.Pp
In regular mode 
.Nm vgrind
accepts input files, processes them, and passes them to 
.Xr groff 1
for output.  
.Pp
In both modes 
.Nm vgrind
passes any lines beginning with a decimal point without conversion.
.Pp
The options are:
.Bl -tag -width Ar
.It Fl 
forces input to be taken from standard input (default if
.Fl f
is specified )
.It Fl T Ar dev
prepares output for the specified device (the default device is 
the default
.Xr groff
output device).
.It Fl d Ar file
specifies an alternate language definitions
file (default is
.Pa /usr/share/misc/vgrindefs )
.It Fl f
forces filter mode
.It Fl h Ar header
specifies a particular header to put on every output page (default is
the file name)
.It Fl l
specifies the language to use.  Currently known are
.Tn PASCAL
.Pq Fl l Ns Ar p ,
.Tn MODEL
.Pq Fl l Ns Ar m ,
C
.Pf ( Fl l Ns Ar c
or the default),
.Tn CSH
.Pq Fl l Ns Ar csh ,
.Tn SHELL
.Pq Fl l Ns Ar sh ,
.Tn RATFOR
.Pq Fl l Ns Ar r ,
.Tn MODULA2
.Pq Fl l Ns Ar mod2 ,
.Tn YACC
.Pq Fl l Ns Ar yacc ,
.Tn LISP
.Pq Fl l Ns Ar isp ,
and
.Tn ICON
.Pq Fl l Ns Ar I .
.It Fl n
forces no keyword bolding
.It Fl s
specifies a point size to use on output (exactly the same as the argument
of a .ps)
.It Fl x
outputs the index file in a ``pretty'' format. 
The index file itself is produced whenever 
.Nm vgrind
is run with a file called 
.Pa index
in the current directory.
The index of function
definitions can then be run off by giving 
.Nm vgrind
the
.Fl x
option and the file
.Pa index
as argument.
.El
.Sh FILES
.Bl -tag -width /usr/share/misc/vgrindefsxx -compact
.It Pa index
file where source for index is created
.It Pa /usr/share/tmac/tmac.vgrind
macro package
.It Pa /usr/libexec/vfontedpr
preprocessor
.It Pa /usr/share/misc/vgrindefs
language descriptions
.El
.Sh SEE ALSO
.Xr vlp 1 ,
.Xr groff 1 ,
.Xr troff 1 ,
.Xr vgrindefs 5
.Sh BUGS
Vfontedpr assumes that a certain programming style is followed:
.Pp
For 
.Tn C
\- function names can be preceded on a line only by spaces, tabs, or an
asterisk.  The parenthesized arguments must also be on the same line.
.Pp
For
.Tn PASCAL
\- function names need to appear on the same line as the keywords
.Em function
or
.Em procedure .
.Pp
For
.Tn MODEL
\- function names need to appear on the same line as the keywords
.Em is beginproc .
.Pp
If these conventions are not followed, the indexing and marginal function
name comment mechanisms will fail.
.Pp
More generally, arbitrary formatting styles for programs mostly look bad.
The use of spaces to align source code fails miserably; if you plan to
.Nm vgrind
your program you should use tabs.  This is somewhat inevitable since the
font used by
.Nm vgrind
is variable width.
.Pp
The mechanism of
.Xr ctags 1
in recognizing functions should be used here.
.Pp
Filter mode does not work in documents using the
.Fl me
or
.Fl ms
macros.
(So what use is it anyway?)
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .


usr.bin/vgrind/vgrind.sh
#!/bin/csh -f
#
#	BSDI	$Id: vgrind.sh,v 1.2 1994/01/03 20:48:12 polk Exp $
#
# Copyright (c) 1980 The Regents of the University of California.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. All advertising materials mentioning features or use of this software
#    must display the following acknowledgement:
#	This product includes software developed by the University of
#	California, Berkeley and its contributors.
# 4. Neither the name of the University nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
#	@(#)vgrind.sh	5.7 (Berkeley) 4/16/91
#

set voptions=
set options=
set files=
set f=''
set head=""
set vf=/usr/libexec/vfontedpr
set tm=/usr/share/tmac
top:
if ($#argv > 0) then
    switch ($1:q)

    case -f:
	set f='filter'
	set options = "$options $1:q"
	shift
	goto top

    case -T*:
	if ($1:q == -T) then
		if ($#argv < 2) then
	    		echo "vgrind: $1:q option must have argument"
	    		goto done
		else
			set voptions=($1:q $2)
			shift
		endif
	else
		set voptions="$voptions $1:q"
	endif
	shift
	goto top

    case -o*:
	set voptions="$voptions $1:q"
	shift
	goto top

    case -d:
	if ($#argv < 2) then
	    echo "vgrind: $1:q option must have argument"
	    goto done
	else
	    set options = ($options $1:q $2)
	    shift
	    shift
	    goto top
	endif
			
    case -h:
	if ($#argv < 2) then
	    echo "vgrind: $1:q option must have argument"
	    goto done
	else
	    set head="$2"
	    shift
	    shift
	    goto top
	endif
			
    case -*:
	set options = "$options $1:q"
	shift
	goto top

    default:
	set files = "$files $1:q"
	shift
	goto top
    endsw
endif
if (-r index) then
    echo > nindex
    foreach i ($files)
	#	make up a sed delete command for filenames
	#	being careful about slashes.
	echo "? $i ?d" | sed -e "s:/:\\/:g" -e "s:?:/:g" >> nindex
    end
    sed -f nindex index >xindex
    if ($f == 'filter') then
	if ("$head" != "") then
	    $vf $options -h "$head" $files | cat $tm/tmac.vgrind -
	else
	    $vf $options $files | cat $tm/tmac.vgrind -
	endif
    else
	if ("$head" != "") then
	    $vf $options -h "$head" $files | \
		sh -c "groff -rx1 $voptions -i -mvgrind 2>> xindex"
	else
	    $vf $options $files | \
		sh -c "groff -rx1 $voptions -i -mvgrind 2>> xindex"
	endif
    endif
    sort -df +0 -2 xindex >index
    rm nindex xindex
else
    if ($f == 'filter') then
	if ("$head" != "") then
	    $vf $options -h "$head" $files | cat $tm/tmac.vgrind -
	else
	    $vf $options $files | cat $tm/tmac.vgrind -
	endif
    else
	if ("$head" != "") then
	    $vf $options -h "$head" $files | groff -i $voptions -mvgrind
	else
	    $vf $options $files | groff -i $voptions -mvgrind
	endif
    endif
endif

done:
